app/Platform/Actions/UpdateGameAchievementsMetricsAction.php

```diff
@@ -65,14 +86,45 @@ public function execute(Game $game): void
             );
             $pointsWeightedTotal += $pointsWeighted;
 
+            // Round percentages to 9 decimal places to match the exact database column precision (decimal(10,9)).
+            // This prevents unnecessary updates due to precision differences in PHP.
+            $unlockPercentage = round($playersTotal ? $unlocksCount / $playersTotal : 0, 9);
+            $unlockHardcorePercentage = round($playersHardcore ? $unlocksHardcoreCount / $playersHardcore : 0, 9);
+
             $achievement->unlocks_total = $unlocksCount;
             $achievement->unlocks_hardcore_total = $unlocksHardcoreCount;
-            $achievement->unlock_percentage = $playersTotal ? $unlocksCount / $playersTotal : 0;
-            $achievement->unlock_hardcore_percentage = $playersHardcore ? $unlocksHardcoreCount / $playersHardcore : 0;
+            $achievement->unlock_percentage = $unlockPercentage;
+            $achievement->unlock_hardcore_percentage = $unlockHardcorePercentage;
             $achievement->TrueRatio = $pointsWeighted;
 
-            $achievement->saveQuietly();
-            $searchIndexingService->queueAchievementForIndexing($achievement->ID);
+            // Only update the achievement if values have actually changed.
+            $isDirty =
+                $achievement->unlocks_total !== $achievement->getOriginal('unlocks_total')
+                || $achievement->unlocks_hardcore_total !== $achievement->getOriginal('unlocks_hardcore_total')
+                || (float) $achievement->unlock_percentage !== (float) $achievement->getOriginal('unlock_percentage')
+                || (float) $achievement->unlock_hardcore_percentage !== (float) $achievement->getOriginal('unlock_hardcore_percentage')
+                || $achievement->TrueRatio !== $achievement->getOriginal('TrueRatio');
+
+            // If the achievement is truly dirty, add it to our list of batch updates.
+            if ($isDirty) {
```

Comment:
A much simpler way to do this is to add casts to the AchievementModel:
```
    protected $casts = [
        ...
        'unlock_percentage' => 'double',
        'unlock_hardcore_percentage' => 'double',
    ];
```
And then use the `$achievement->isDirty()` method. I was doing that earlier when we were discussing the issue.